### PR TITLE
feat: Add "deno ast" subcommand

### LINKED ISSUE #2355: "deno ast script.ts"
Should print to stdout a JSON blob representing the typescript AST.

### PR DESCRIPTION
**Pull request by littledivy**
Leverages the swc ast parser utility behind the `--unstable` flag

Closes #2355

### CODE REVIEW

File: cli/main.rs [hunk 1]
```diff
@@ -436,6 +437,35 @@ async fn eval_command(
   Ok(())
 }
 
+async fn ast_command(flags: Flags, source_file: String) -> Result<(), ErrBox> {
+  let module_specifier = ModuleSpecifier::resolve_url_or_path(&source_file)?;
+
+  debug!(">>>>> ast START");
+  let global_state = GlobalState::new(flags)?;
+
+  info!("{} {}", colors::green("AST"), module_specifier.to_string());
```
Comment (bartlomieju) [inline cli/main.rs:446]: Not needed

File: cli/main.rs [hunk 2]
```diff
@@ -436,6 +437,35 @@ async fn eval_command(
...
+
```
Comment (bartlomieju) [inline cli/main.rs:447]: Missing `--unstable` check

File: cli/main.rs [hunk 3]
```diff
@@ -436,6 +437,35 @@ async fn eval_command(
...
+    out.media_type,
+    src,
+    |parse_result| {
+      let module = parse_result.unwrap();
+      Ok(module)
+    },
+  )?;
+
+  debug!(">>>>> ast END");
+  println!("{}", serde_json::to_string(&ast)?);
```
Comment (bartlomieju) [inline cli/main.rs:465]: Use `write_to_stdout_ignore_sigpipe` instead

File: cli/flags.rs [hunk 4]
```diff
@@ -731,10 +744,8 @@ The installation root is determined, in order of precedence:
 These must be added to the path manually if required.")
 }
 
-fn bundle_subcommand<'a, 'b>() -> App<'a, 'b> {
-  SubCommand::with_name("bundle")
-    .arg(lock_arg())
-    .arg(lock_write_arg())
+fn ast_subcommand<'a, 'b>() -> App<'a, 'b> {
+  SubCommand::with_name("ast")
```
Comment (bartlomieju) [inline cli/flags.rs:748]: This is wrong, somehow edited `bundle_subcommand` instead of adding new subcommand

File: cli/flags.rs [hunk 5]
```diff
@@ -755,6 +766,21 @@ If no output file is given, the output is written to standard output:
     )
 }
 
+fn bundle_subcommand<'a, 'b>() -> App<'a, 'b> {
```
Comment (bartlomieju) [inline cli/flags.rs:769]: Ditto, but the other way around

File: cli/flags.rs (hunk 4 above)
Comment (divy-work) [reply cli/flags.rs:748]: Oh, no idea how that happened!

Comment (divy-work) [reply cli/flags.rs]: Fixed

**Comment by divy-work** [comment]
Sure will do. I think if future we can add an options to write the output json to a file.

**Comment by narwy** [comment]
I think we need the `--unstable` to show that the output currently isn't stable, which I think @ry was getting at?

**Comment by divy-work** [comment]
> I think we need the `--unstable` to show that the output currently isn't stable, which I think @ry was getting at?

Yes, it is currently behind `--unstable` but the output is quite stable and also has strong typings available. 
See https://github.com/swc-project/swc/blob/master/node-swc/src/types.ts

File: cli/tests/integration_tests.rs [hunk 6]
```diff
@@ -694,16 +694,152 @@ fn ts_reload() {
 
 #[test]
 fn test_ast() {
-  let mod1 = util::root_path().join("cli/tests/subdir/mod1.ts");
+  let mod1 = util::root_path().join("cli/tests/subdir/print_hello.ts");
   assert!(mod1.is_file());
-  let mut deno = util::deno_cmd()
+  let output = util::deno_cmd()
     .current_dir(util::root_path())
     .arg("ast")
```
Comment (bartlomieju) [inline cli/tests/integration_tests.rs]: Why not use `itest` instead? I would be much shorter?

Comment (divy-work) [reply cli/tests/integration_tests.rs]: Right, will do that.

**Comment by divy-work** [comment]
Hope it works now lol. _(2 hours wasted)_

**Comment by bartlomieju** [comment]
https://github.com/denoland/deno/issues/2355#issuecomment-719522155

**Pull request closed without merging**